rollup/fees/rollup_fee.go

```diff
@@ -138,3 +140,66 @@ func mulAndScale(x *big.Int, y *big.Int, precision *big.Int) *big.Int {
 	z := new(big.Int).Mul(x, y)
 	return new(big.Int).Quo(z, precision)
 }
+
+// copyTransaction copies the transaction, removing the signature
+func copyTransaction(tx *types.Transaction) *types.Transaction {
+	if tx.To() == nil {
+		return types.NewContractCreation(
+			tx.Nonce(),
+			tx.Value(),
+			tx.Gas(),
+			tx.GasPrice(),
+			tx.Data(),
+		)
+	}
+	return types.NewTransaction(
+		tx.Nonce(),
+		*tx.To(),
+		tx.Value(),
+		tx.Gas(),
+		tx.GasPrice(),
+		tx.Data(),
+	)
+}
+
+func CalculateTotalFee(tx *types.Transaction, state StateDB) (*big.Int, error) {
+	unsigned := copyTransaction(tx)
+	raw, err := rlpEncode(unsigned)
+	if err != nil {
+		return nil, err
+	}
+
+	l1BaseFee, overhead, scalar := readGPOStorageSlots(rcfg.L1GasPriceOracleAddress, state)
+	l1Fee := CalculateL1Fee(raw, overhead, l1BaseFee, scalar)
+
+	l2GasLimit := new(big.Int).SetUint64(tx.Gas())
+	l2Fee := new(big.Int).Mul(tx.GasPrice(), l2GasLimit)
+	fee := new(big.Int).Add(l1Fee, l2Fee)
+	return fee, nil
+}
+
+func VerifyFee(signer types.Signer, tx *types.Transaction, state StateDB) error {
+	fee, err := CalculateTotalFee(tx, state)
+	if err != nil {
+		return fmt.Errorf("invalid transaction: %w", err)
+	}
+
+	// Prevent transactions without enough balance from
+	// being accepted by the chain but allow through 0
+	// gas price transactions
+	cost := tx.Value()
+	if tx.GasPrice().Cmp(common.Big0) != 0 {
```

Comment:
Why only add `fee` when `tx.gasPrice > 0`?

Reply:
It's the same in optimism.

I think they don't change L1&L2 if gasPrice=0.  (However, only `GasPriceOracle` is allowed to send 0 gasPrice tx)

Reply:
I will remove this condition now